### PR TITLE
GTK 3.20 :: Fixed #486

### LINKED ISSUE #486: Titlebar height
Window titlebar height in [Terminix](https://github.com/gnunn1/terminix) is biggest than in other apps when I use [Numix](https://github.com/numixproject/numix-gtk-theme) theme

### PR DESCRIPTION
**Pull request by megax**
Fixes #486

### CODE REVIEW

**Comment by megax** [comment]
@khurshid-alam Rebased code and modified. Please test.

**Comment by khurshid-alam** [comment]
@megax Merge conflict. Please rebase. Then only I can test.

**Comment by megax** [comment]
@khurshid-alam Rebased.

File: gtk-3.20/scss/widgets/_toolbar.scss [hunk 1]
```diff
@@ -33,86 +32,94 @@
...
-
-    .titlebar {
+    %titlebar { // Default headerbar and titlebar code.
+        @include toolbar($titlebar_bg_color, $titlebar_fg_color);
         @include linear-gradient($titlebar_bg_color);
 
         border-radius: $roundness $roundness 0 0;
         color: mix($titlebar_fg_color, $titlebar_bg_color, .1);
+        padding: 0 $spacing;
+        min-height: 42px;
```
Comment (khurshid-alam) [inline gtk-3.20/scss/widgets/_toolbar.scss:122]: @megax  Why 42px? remove it.

File: gtk-3.20/scss/widgets/_toolbar.scss [hunk 2]
```diff
@@ -33,86 +32,94 @@
...
         border-style: solid;
 
+        // add vertical margins to common widget on the headerbar to avoid them spanning the whole height
+        entry,
+        spinbutton,
+        separator,
+        switch,
+        button { // Size height
+            margin-top: $spacing + 2px;
+            margin-bottom: $spacing + 2px;
```
Comment (khurshid-alam) [inline gtk-3.20/scss/widgets/_toolbar.scss]: @megax Use `$spacing + 3px` for margins

File: gtk-3.20/scss/widgets/_toolbar.scss [hunk 3]
```diff
@@ -121,24 +128,32 @@
             text-shadow: none;
         }
 
-        &.default-decoration {
-            border: 0;
+        &.default-decoration { // Default titlebar (old metacity)
+            min-height: 24px;
```
Comment (khurshid-alam) [inline gtk-3.20/scss/widgets/_toolbar.scss:132]: @megax `min-height: 24px` ? Why is it necessary?

File: gtk-3.20/scss/widgets/_toolbar.scss [hunk 4]
```diff
@@ -121,24 +128,32 @@
...
             box-shadow: none;
+            border: 0;
+
+            button.titlebutton {
+                min-height: 22px;
+                min-width: 22px;
```
Comment (khurshid-alam) [inline gtk-3.20/scss/widgets/_toolbar.scss]: min-height: 22px;  min-width: 22px; again why?

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 1 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss:122]: Not, to be awarded https://github.com/horst3180/arc-theme/blob/master/common/gtk-3.0/3.20/sass/_common.scss#L962

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 2 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss]: Sets the height of buttons, i think so regular, commissioner rewrite?

image
image

Example: `$spacing + 1px`
image

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 3 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss:132]: default window titlebar size (gnome-shell), size of a (metacity, unity...):
image

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 4 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss]: default window titlebar size (gnome-shell), size of a (metacity, unity...) -> titlebutton size [X]:
image

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 3 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss:132]: If you are not:

image

**Comment by megax** [comment]
@khurshid-alam Unity does not use these. I tested three environment but by the time I put this together so... :)

cinnamon, gnome-shell, unity

**Comment by megax** [comment]
gnome-shell:
image

cinnamon:
image

unity:
image

**Comment by megax** [comment]
1px bug -> other font: ubunut, font size... I have not figured out why there =/

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 2 above)
Comment (khurshid-alam) [reply gtk-3.20/scss/widgets/_toolbar.scss]: 2px is too less, I wanted to use 

``` css
        button { // Size height
            margin-top: $spacing + 3px;
            margin-bottom: $spacing + 3px;
        }
```

It looks nice.

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 4 above)
Comment (khurshid-alam) [reply gtk-3.20/scss/widgets/_toolbar.scss]: @megax Then shouldn't that be 16x16 ?

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 2 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss]: Okay :)

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 4 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss]: I try.

**Comment by megax** [comment]
@lots0logs Please test button, now is a good principle.

image

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 4 above)
Comment (megax) [reply gtk-3.20/scss/widgets/_toolbar.scss]: @khurshid-alam modified 16x16

**Comment by khurshid-alam** [comment]
@megax 

## Potential regression (Unity):

Issue #290 : Window title is 1px wider than window itself
unity-border

Which I fixed it in 06197571ed128220bfe33f50c4ebfd6e02b0a532 for 3.18

For 3.20 if I apply same fix after merging this, It fixes the issue but there remains a white dot between Unity titlebar and window.

white-dot-in-window

**Comment by megax** [comment]
@khurshid-alam i looking for a solution.

File: gtk-3.20/scss/widgets/_toolbar.scss (hunk 1 above)
Comment (khurshid-alam) [reply gtk-3.20/scss/widgets/_toolbar.scss:122]: I don't think it's necessary. I didn't notice any different without it.

**Comment by megax** [comment]
@khurshid-alam LOL, headerbar radius = px bug...
image

```
    headerbar {
        border-radius: 0;

        window:not(.tiled):not(.maximized) separator:first-child + &, // tackles the paned container case
        window:not(.tiled):not(.maximized) &:first-child { &:backdrop, & { border-top-left-radius: 0; } }

        window:not(.tiled):not(.maximized) &:last-child { &:backdrop, & { border-top-right-radius: 0; } }
}
```

to be placed Unitydecoration (have not figured out that). Idea?
image

**Comment by khurshid-alam** [comment]
> ...to be placed Unitydecoration (have not figured out that). Idea?

No. I don't think there is any way to do that. https://wiki.ubuntu.com/Unity/Theming

**Pull request merged**